src/components/DataCollector/index.js

```diff
@@ -20,20 +20,28 @@ const createDataCollector = () => {
   let network;
 
   const makeServerCall = event => {
-    const { expectsResponse } = event;
     const payload = network.createPayload();
     payload.addEvent(event);
-    return network.sendRequest(payload, expectsResponse).then(response => {
-      const data = {
-        requestBody: clone(payload)
-      };
+    const responsePromise = Promise.resolve()
+      .then(() => {
+        return lifecycle.onBeforeDataCollection(payload, responsePromise);
+      })
+      .then(() => {
+        return network
+          .sendRequest(payload, payload.expectsResponse)
+          .then(response => {
+            const data = {
+              requestBody: clone(payload)
+            };
 
-      if (response) {
-        data.responseBody = clone(response);
-      }
+            if (response) {
+              data.responseBody = clone(response);
+            }
 
-      return data;
-    });
+            return data;
+          });
+      });
```

Comment:
I recommend keeping promise chaining as shallow when possible rather than nesting chains. For example:

```
    const responsePromise = Promise.resolve()
      .then(() => {
        return lifecycle.onBeforeDataCollection(payload, responsePromise);
      })
      .then(() => {
        return network.sendRequest(payload, payload.expectsResponse);
      })
      .then(response => {
          const data = {
            requestBody: clone(payload)
          };

           if (response) {
            data.responseBody = clone(response);
          }

          return data;
      });
```

Reply:
Yes, that will make it less confusing.